### ISSUE TITLE
Performance issues while rendering a lot of q-input components

### ISSUE THREAD

**Issue by Codebreaker101** [quasarframework/quasar#12369]
### What happened?

When having a lot of q-input components on a page it takes a while to render, especially if you have label and hint enabled.

Here are some screenshots from vue dev tools:
![Screenshot from 2022-02-06 15-58-29](https://user-images.githubusercontent.com/16204269/152686864-5ffe3708-b2eb-4334-973f-b73df34b926c.png)
![Screenshot from 2022-02-06 15-59-04](https://user-images.githubusercontent.com/16204269/152686866-6bf72ca5-b97b-463e-b266-328bdc0bca8f.png)

Is this how it is supposed to be? If yes, what can one do to improve performance while rendering lots of inputs? 

### What did you expect to happen?

I expected for it not to be that slow.

### Reproduction URL

https://codesandbox.io/s/musing-glade-nlkdp?file=/src/pages/Index.vue

### How to reproduce?

1. Go to provided link
2. Click on TRIGGER
3. Depending on you machine you can increase the number of components rendered

### Flavour

Quasar CLI (@quasar/cli | @quasar/app)

### Areas

Components (quasar)

### Platforms/Browsers

Firefox, Chrome

### Quasar info output

```shell
Operating System - Linux(5.10.93-1-MANJARO) - linux/x64
NodeJs - 14.17.6

Global packages
  NPM - 6.14.15
  yarn - 1.22.17
  @quasar/cli - 1.2.1
  @quasar/icongenie - 2.4.0
  cordova - Not installed

Important local packages
  quasar - 2.5.1 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app - 3.3.2 -- Quasar Framework local CLI
  @quasar/extras - 1.12.4 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.2.29 -- The progressive JavaScript framework for building modern web UI.
  vue-router - 4.0.12
  vuex - 4.0.2 -- state management for Vue.js
  electron - 17.0.0 -- Build cross platform desktop apps with JavaScript, HTML, and CSS
  electron-packager - 15.4.0 -- Customize and package your Electron app with OS-specific bundles (.app, .exe, etc.) via JS or CLI
  electron-builder - Not installed
  @babel/core - 7.16.12 -- Babel compiler core.
  webpack - 5.68.0 -- Packs CommonJs/AMD modules for the browser. Allows to split your codebase into multiple bundles, which can be loaded on demand. Support loaders to preprocess files, i.e. json, jsx, es7, css, less, ... and your custom stuff.
  webpack-dev-server - 4.7.3 -- Serves a webpack app. Updates the browser on changes.
  workbox-webpack-plugin - Not installed
  register-service-worker - 1.7.2 -- Script for registering service worker, with hooks
  typescript - 4.5.5 -- TypeScript is a language for application scale JavaScript development
  @capacitor/core - 3.2.4 -- Capacitor: Cross-platform apps with JavaScript and the web
  @capacitor/cli - 3.2.4 -- Capacitor: Cross-platform apps with JavaScript and the web
  @capacitor/android - 3.2.4 -- Capacitor: Cross-platform apps with JavaScript and the web
  @capacitor/ios - Not installed

Quasar App Extensions
  *None installed*
```

**Comment by cohlar** [comment]
+1
I use Quasar with Vite, I also have pages with a lot of inputs and it makes my app ridiculously slow.
Would love some performance tips.

**Comment by khanhnm222** [comment]
I have used both Vite and Webpack build modes for my project, which renders a lot of q-table/q-btn,... components on one page, and my app loads very slowly. 
Hopefully, there will be a solution to improve performance.

**Comment by azizaza** [comment]
same here! any solution? 
i'm using quasar with vite

**Comment by AlexDaniel** [comment]
Ugh, the only solution I know (and use) is to create a `progressiveLoading` ref which is just a counter.

Then you do:
```js
function bumpProgressiveLoading() {
  progressiveLoading.value += 1;
  const maxLength = myArrayWithElements.length;
  if (progressiveLoading.value < maxLength - 1) {
    nextTick(() => {
      setTimeout(bumpProgressiveLoading, 1);
    });
  }
}
onMounted(bumpProgressiveLoading);
```
Then, in your template, add a v-if/v-else based on the index. Render a `q-skeleton` if the index didn't reach the element's index. `q-skeletons` are very fast to render compared to `q-inputs`.

This means that as the user navigates to your page, they will just see a list of skeletons first, and that list will progressively turn into something useful. Pay attention to rerenders and ensure that your q-inputs don't get rerendered every time (you will probably need to create components for each item in the array even if you don't want to).

Basically, it's manual time slicing.

**Comment by AlexDaniel** [comment]
Actually, this makes me think that Vue should provide a helper for time-slicing… This way, you could batch many updates into a single render cycle while maintaining responsiveness.

That being said, still, why the hell are q-inputs so slow to render?

**Comment by rstoenescu** [comment]
Closing this with a full write-up, since the thread asked "why are q-inputs so slow to render" and deserves a real answer.

**Not a bug.** Measured today on the built production bundles (Playwright, Chromium, 2000 QInputs with a label and a hint on one page): a QInput costs about 75 µs of JavaScript to mount and about 100 µs until it is painted. 2000 of them take ~150 ms of JS and ~200 ms to first paint. Roughly a third of that is the browser laying out and painting the markup itself (the same 18k DOM nodes as static HTML cost ~67 ms); the rest is Vue creating 2000 component instances with ~60 props, a few watchers and computeds each. A keystroke afterwards costs ~4 ms for all 2000 fields, so re-rendering is not a factor either. There is no single hot spot inside QInput. It is an ordinary component doing an ordinary amount of work, and 500 or 1000 of them on one page cost what 500 or 1000 stateful components cost in any framework.

**What the screenshots showed.** The timelines in the report are the Vue devtools performance tab in development mode. Dev builds of Vue and Quasar are 50-55% slower than production here, and the devtools instrumentation adds its own overhead on top, so those bars overstate what users see by a wide margin. The one real thing the second screenshot shows is that about half of each "QInput mount" bar is "Transition mount / BaseTransition": the animation wrapper around the hint/error message.

**Then and now.** Quasar 2.5.1 (the version in the report) and 2.30.1 mount a QInput at the same speed within noise (+2 to 6%). The Vue 3.2 to 3.5 upgrade costs more than that (+18%), mostly because Vue 3.5's BaseTransition became heavier per instance: that message Transition went from ~9 µs per field on Vue 3.2 to ~18 µs on Vue 3.5, roughly a quarter of a QInput's mount.

**Why we still made a perf commit.** Nothing was broken, but that Transition was doing very little for its price: with no `appear`, it only animated the hint-to-error swap, and the messages element is already keyed by its content, so Vue re-creates it on a swap anyway. 7bf2244cf replaces the Transition with a CSS keyframe on the re-created element, applied only after a swap has happened past the first render, so nothing animates on page load and no field gets a composited layer at mount. The enter animation is the same (fade in from -10px, same curve and duration). What changed: the old message no longer fades out in place while the new one fades in; it is removed at once. The `q-transition--field-message-*` classes no longer exist; anyone restyling the swap through them should target `.q-field__messages--animated` instead. Result: 2000 QInputs with hints mount 25% faster (201 ms to 150 ms of JS in the same run), and every form with hints or validation rules benefits. Ships with the next release.

**If you have hundreds of fields on one page**, that is a virtualisation or progressive-rendering problem no component library can remove on its own: use `QVirtualScroll`, render sections on demand with `QIntersection`, or time-slice as described above. `hide-bottom-space` also skips the message animation entirely if you do not need it.
